File: backend-example.js
// Backend Example: Node.js with Gemini API Integration
// This shows how to integrate with Google's Gemini API for real image processing

const { GoogleGenerativeAI } = require("@google/generative-ai")
const fs = require("fs")
const express = require("express")
const app = express()

app.use(express.json({ limit: "10mb" }))

// Initialize Gemini API
const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY)

async function processImageWithGemini(base64Image, emotion, hallucinationEnabled) {
  try {
    // Get the generative model
    const model = genAI.getGenerativeModel({ model: "gemini-pro-vision" })

    // Convert base64 to buffer
    const imageBuffer = Buffer.from(base64Image.split(",")[1], "base64")

    // Create emotion prompt based on settings
    const intensity = hallucinationEnabled ? "extreme and exaggerated" : "subtle and realistic"
    const emotionDescriptions = {
      "-10": "apocalyptic despair and cosmic horror",
      "-8": "extreme terror and overwhelming fear",
      "-6": "devastating sorrow and deep anguish",
      "-4": "intense anger and fury",
      "-2": "sadness and disappointment",
      "-1": "slight sadness or melancholy",
      0: "neutral and calm expression",
      1: "happiness and contentment",
      2: "joy and excitement",
      4: "euphoric joy and elation",
      6: "transcendent bliss and wonder",
      8: "cosmic ecstasy and divine joy",
      10: "divine rapture and ultimate bliss",
    }

    const emotionDesc = emotionDescriptions[emotion.toString()] || "neutral expression"

    const prompt = `Transform this portrait image to show ${emotionDesc}. 
    Apply ${intensity} emotional transformation while maintaining the person's identity and basic facial structure. 
    Focus on facial expression, eye emotion, mouth position, and overall emotional aura. 
    ${hallucinationEnabled ? "Feel free to add dramatic lighting, color shifts, or surreal elements that enhance the emotion." : "Keep the transformation natural and believable."}
    Return only the transformed image.`

    // Prepare image data for Gemini
    const imagePart = {
      inlineData: {
        data: imageBuffer.toString("base64"),
        mimeType: "image/jpeg",
      },
    }

    // Generate content with image and prompt
    const result = await model.generateContent([prompt, imagePart])
    const response = await result.response

    // Note: Gemini Pro Vision currently returns text descriptions, not images
    // For actual image generation, you might need to use:
    // 1. Gemini + another image generation service
    // 2. Google's Imagen API (when available)
    // 3. Other AI services like Replicate, Stability AI, etc.

    // For now, return the description and use it with an image generation service
    const description = response.text()

    return {
      success: true,
      description: description,
      // In real implementation, you'd generate image from description
      processedImageUrl: await generateImageFromDescription(description),
      metadata: {
        model: "gemini-pro-vision",
        emotion: emotion,
        hallucinationEnabled: hallucinationEnabled,
        prompt: prompt,
      },
    }
  } catch (error) {
    console.error("Gemini API Error:", error)
    throw new Error(`Failed to process image with Gemini: ${error.message}`)
  }
}

// Example function to generate image from Gemini's description
async function generateImageFromDescription(description) {
  // This would integrate with an image generation service
  // Examples: Replicate, Stability AI, DALL-E, Midjourney API, etc.

  // Placeholder implementation
  return "data:image/jpeg;base64,..." // Generated image as base64
}

// Express.js route handler example
app.post("/api/update-image", async (req, res) => {
  try {
    const { emotion, image, hallucinationEnabled } = req.body

    // Validate inputs
    if (!image || emotion === undefined) {
      return res.status(400).json({ error: "Missing required fields" })
    }

    // Process with Gemini
    const result = await processImageWithGemini(image, emotion, hallucinationEnabled)

    res.json({
      processedImage: result.processedImageUrl,
      description: result.description,
      metadata: result.metadata,
    })
  } catch (error) {
    console.error("API Error:", error)
    res.status(500).json({ error: "Failed to process image" })
  }
})

// Environment variables needed:
// GEMINI_API_KEY=your_gemini_api_key_here

File: backend-example.py
# Backend Example: Python with Gemini API Integration
# Alternative implementation using Python and Google's Generative AI library

import google.generativeai as genai
import base64
import io
from PIL import Image
import os
from flask import Flask, request, jsonify

# Configure Gemini API
genai.configure(api_key=os.environ['GEMINI_API_KEY'])

app = Flask(__name__)

def process_image_with_gemini(base64_image, emotion, hallucination_enabled):
    """
    Process image using Gemini Pro Vision API
    """
    try:
        # Initialize the model
        model = genai.GenerativeModel('gemini-pro-vision')
        
        # Convert base64 to PIL Image
        image_data = base64_image.split(',')[1]
        image_bytes = base64.b64decode(image_data)
        image = Image.open(io.BytesIO(image_bytes))
        
        # Create emotion-based prompt
        intensity = "extreme and exaggerated" if hallucination_enabled else "subtle and realistic"
        
        emotion_descriptions = {
            -10: "apocalyptic despair and cosmic horror",
            -8: "extreme terror and overwhelming fear", 
            -6: "devastating sorrow and deep anguish",
            -4: "intense anger and fury",
            -2: "sadness and disappointment",
            -1: "slight sadness or melancholy",
            0: "neutral and calm expression",
            1: "happiness and contentment", 
            2: "joy and excitement",
            4: "euphoric joy and elation",
            6: "transcendent bliss and wonder",
            8: "cosmic ecstasy and divine joy",
            10: "divine rapture and ultimate bliss"
        }
        
        emotion_desc = emotion_descriptions.get(emotion, "neutral expression")
        
        prompt = f"""Transform this portrait image to show {emotion_desc}. 
        Apply {intensity} emotional transformation while maintaining the person's identity and basic facial structure.
        Focus on facial expression, eye emotion, mouth position, and overall emotional aura.
        {'Feel free to add dramatic lighting, color shifts, or surreal elements that enhance the emotion.' if hallucination_enabled else 'Keep the transformation natural and believable.'}
        Describe the transformed image in detail."""
        
        # Generate content
        response = model.generate_content([prompt, image])
        description = response.text
        
        # In a real implementation, you would:
        # 1. Use the description to generate an actual image
        # 2. Integrate with image generation APIs (Replicate, Stability AI, etc.)
        # 3. Return the generated image
        
        return {
            'success': True,
            'description': description,
            'processed_image_url': generate_image_from_description(description),
            'metadata': {
                'model': 'gemini-pro-vision',
                'emotion': emotion,
                'hallucination_enabled': hallucination_enabled,
                'prompt': prompt
            }
        }
        
    except Exception as e:
        print(f"Gemini API Error: {e}")
        raise Exception(f"Failed to process image with Gemini: {str(e)}")

def generate_image_from_description(description):
    """
    Generate image from Gemini's description using an image generation service
    This is a placeholder - integrate with your preferred image generation API
    """
    # Example integrations:
    # - Replicate API
    # - Stability AI
    # - OpenAI DALL-E
    # - Midjourney API
    
    # Placeholder return
    return "data:image/jpeg;base64,..."  # Generated image as base64

@app.route('/api/update-image', methods=['POST'])
def update_image():
    try:
        data = request.get_json()
        emotion = data.get('emotion')
        image = data.get('image') 
        hallucination_enabled = data.get('hallucinationEnabled', False)
        
        # Validate inputs
        if not image or emotion is None:
            return jsonify({'error': 'Missing required fields'}), 400
            
        # Process with Gemini
        result = process_image_with_gemini(image, emotion, hallucination_enabled)
        
        return jsonify({
            'processedImage': result['processed_image_url'],
            'description': result['description'], 
            'metadata': result['metadata']
        })
        
    except Exception as e:
        print(f"API Error: {e}")
        return jsonify({'error': 'Failed to process image'}), 500

if __name__ == '__main__':
    app.run(debug=True)

# Environment variables needed:
# GEMINI_API_KEY=your_gemini_api_key_here
